ksql-rest-app/src/main/java/io/confluent/ksql/rest/server/HeartbeatHandler.java

```diff
@@ -0,0 +1,341 @@
+        HostInfo host = entry.getKey();
+        TreeMap<Long, HeartbeatInfo> heartbeats = entry.getValue();
+        //Question: Until this thread gets the lock, "now" will have changed. But if I compute "now"
+        //  inside synchronized, then different hosts will have different windows
+        synchronized (heartbeats) {
+          // 1. remove heartbeats older than window
+          while (!heartbeats.isEmpty()) {
+            Entry<Long, HeartbeatInfo> nestedEntry = heartbeats.firstEntry();
+            long timestamp = nestedEntry.getKey();
+            if (timestamp < windowStart) {
+              heartbeats.pollFirstEntry();
+            } else {
+              break;
+            }
+          }
+          // 2. count consecutive missed heartbeats and mark as up
+          downPolicy(host, windowStart, windowEnd, heartbeats);
+        }
+      }
+    }
+
+
+    private void downPolicy(final HostInfo host, final long windowStart, final long windowEnd,
+                            final TreeMap<Long, HeartbeatInfo> heartbeats) {
+      long missedCount = 0;
+      long prev = windowStart;
+      // No heartbeat received in this window
+      if (heartbeats.isEmpty()) {
+        aliveHosts.put(host.toString(), false);
+        return;
+      }
+      for (long ts : heartbeats.keySet()) {
+
+        //Don't count heartbeats after window end
+        if (ts >= windowEnd) {
+          continue;
+        }
+        if (ts - heartbeatSendInterval > prev) {
+          missedCount = (ts - prev - 1) / heartbeatSendInterval;
+        } else {
+          missedCount = 0;
+        }
+        prev = ts;
+      }
+      // Check frame from last received heartbeat to window end
+      if (windowEnd - prev - 1 > 0) {
+        missedCount = (windowEnd - prev - 1) / heartbeatSendInterval;
+      }
+
+      if (missedCount < heartbeatMissedThreshold) {
+        aliveHosts.put(host.toString() , true);
+
+      } else {
+        aliveHosts.put(host.toString(), false);
+      }
+
+    }
+  }
+
+  class SendHeartbeat implements Runnable {
+
+    @Override
+    public void run() {
+      try {
+        //broadcast heartbeat
+        remoteHostsURI.forEach((uri, bool) -> serviceContext.getKsqlClient().sendHeartbeatRequest(
+            uri, localHostInfo, System.currentTimeMillis()));
+      } catch (Exception e) {
+        e.printStackTrace();
+      }
+    }
+  }
+
+  class DiscoverCluster implements Runnable {
+
+    @Override
+    public void run() {
+
+      List<PersistentQueryMetadata> currentQueries = engine.getPersistentQueries();
+      if (currentQueries.isEmpty()) {
+        return;
+      }
+
+      Set<HostInfo> uniqueHosts = currentQueries.stream()
+          .map(queryMetadata -> ((QueryMetadata)queryMetadata).getKafkaStreams().allMetadata())
+          .flatMap(Collection::stream)
+          .map(StreamsMetadata::hostInfo)
+          .filter(hostInfo -> !(hostInfo.host().equals(localHostInfo.host())
```

Comment:
general comment.. we should make sure the advertised hostname (one in streams metadata), matches with the `localHostInfo`. Just discuss this once with alan, as the advertised listeners are being reworked now